pkg/watch/replay.go

```diff
@@ -0,0 +1,201 @@
+package watch
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+type replayRequest struct {
+	r        *Registrar
+	gvk      schema.GroupVersionKind
+	isCancel bool // If true, this is a request to cancel a pending replay.
+}
+
+type cancelMap map[*Registrar]context.CancelFunc
+
+func (m *cancelMap) Set(r *Registrar, c context.CancelFunc) {
+	if *m == nil {
+		*m = make(cancelMap)
+	}
+	(*m)[r] = c
+}
+
+// replayEventsLoop processes requests to start and stop replaying events for
+// registrars that join an existing informer and need historical data.
+// Events for each registrar will be listed and replayed in an independent goroutine.
+func (wm *Manager) replayEventsLoop() error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	ctx, cancel := ContextForChannel(wm.stopped)
+	defer cancel()
+
+	// Entries remain until a watch is removed.
+	m := make(map[schema.GroupVersionKind]cancelMap)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case req := <-wm.replayRequests:
+			byRegistrar, inProgress := m[req.gvk]
```

Comment:
Should `inProgress` be triggered by whether the replay is in `byRegistrar`?

What happens if Registrar A adds a watch, Registrar B adds a watch (and gets replay), and then Registrar C adds a watch? Will a replay trigger?

Reply:
You're right, this looks like a bug. I'll fix and add a corresponding test scenario. Good catch!

Reply:
Extended test (which initially failed 👍) and fixed.